### PR TITLE
Ability to Disable bones, drop items or keep inventory.

### PR DESCRIPTION
**Pull request by tenplus1**
I've added a setting to minetest.conf called `bones_mode` that allows the player or admin to enable, disable or drop items when you die.  Setting to `bones` will place bones or drop as items inside protected area, `drop` will drop all items and `keep` will leave your inventory alone.

### CODE REVIEW

**Comment by kilbith** [comment]
@tenplus1 You really should stop duplicating your PRs and learn how to squash your commits.

https://github.com/minetest/minetest_game/pull/915
https://github.com/minetest/minetest_game/pull/914
https://github.com/minetest/minetest_game/pull/912
https://github.com/minetest/minetest_game/pull/896
https://github.com/minetest/minetest_game/pull/740

**Comment by tenplus1** [comment]
I tried to rebase this pull and remove changes already applied with only the latest one's to be added but it was all too complicated and I felt this was the easier and cleanest method.  my bad.

**Comment by sofar** [comment]
Add it in the ```game_api.txt``` since it's an API? Otherwise this looks good.

**Comment by tenplus1** [comment]
I've merged and squashed PR #966 into this one thanks to Sofar.

**Comment by MT-Modder** [comment]
:+1:

**Comment by DerpyPony25** [comment]
1+ for you :3

**Comment by DonBatman** [comment]
:+1:

**Comment by kaeza** [comment]
Other than this should be a `minetest.conf` option, :+1:.

**Comment by tenplus1** [comment]
Having this as a global switch is a lot easier than having a .conf option, it means that if you load DropOnDie mod it automatically disables dropped bones and saves the player editing files.  This feature has been tested on Xanadu and works well.

**Comment by sofar** [comment]
It's not easier for people who don't use mods that disable it. They need to edit the code now.

It should be both, ideally. Both a minetest.conf setting and a switch that mods can use.

**Comment by tenplus1** [comment]
All the mod maker has to do is add a dependency and a single line to their mod:  bones.enable = false

How much easier can it be ?  This way players don't have to do a thing apart from enable the mod.

**Comment by sofar** [comment]
What about players who don't have your mod?

**Comment by tenplus1** [comment]
If a player doesn't use any mods to change item drop when dead then bones will appear by default.

**Comment by sofar** [comment]
So how does one disable bones without any mods?

**Comment by MT-Modder** [comment]
I guess the whole point is for another mod to use it...

But why wouldn't you kill two birds with one stone and just use a .conf setting?

It's still a single line addition: ```minetest.setting_setbool("enable_bones", false)```.
Plus a mod wouldn't need to add a dependency.

Seems kinda silly not to.

**Comment by tenplus1** [comment]
Ya know, I originally added the minetest.conf setting and was told to find an easier way to disable bones so that players wouldn't need to edit a file.  This way it leaves it up to the mod maker instead of the player.

**Comment by MT-Modder** [comment]
Wouldn't it be better to leave it up to the modder **AND** player not just the modder? As a setting it's easy for both.

The only thing easier than changing a true/false setting is doing nothing...so people want the change but don't want to change it themselves. That's just lazy.

"I want to use mapgen v5, but I have to click something to use it, that's too hard!"

**Comment by tenplus1** [comment]
Now has .conf and mod setting to disable bones.

**Comment by MT-Modder** [comment]
Thank you! I'm curious as to why we need both though?

minetest.setting_setbool("enable_bones", false) **vs** bones.enable = false

27 Character difference. Is one faster than the other?
If so I'd keep both but if not then I'd just keep the .conf setting.

**Comment by tenplus1** [comment]
@MT-Modder - The default behaviour of bones mod is to drop bones on death, the .conf setting is to change default behaviour if a player wishes to disable bones and not drop anything, while the bones.enable global allows mod makers to change the setting when their mod is run.

**Comment by MT-Modder** [comment]
Right but the mod maker can just use the .conf setting for the same purpose. That's all I'm saying.

**Comment by sofar** [comment]
Mods can't use the ```mintest.conf``` setting method since that's too late - it wouldn't take effect until the entire server is restarted.

the bones mod reads the bones setting before any custom mod. You can't change that order, either.

**Comment by MT-Modder** [comment]
Oh, my bad. Thanks for clearing that up for me.

**Comment by jhcole** [comment]
:+1: I need this

**Comment by tenplus1** [comment]
@paramat - can we get a thumbs up from you also ?

**Comment by paramat** [comment]
I don't have much opinion on disablng bones so i think other MTGame devs should decide.There are many other PRs and issues on this, what do the other devs think in those?

**Comment by Fixer-007** [comment]
:+1:

**Comment by everamzah** [comment]
I think that if `tnt` and `fire` mods are going to have settings that determine whether they're enabled, then so should `bones`.  In my opinion it should be a simple setting check, as it is in the aforementioned mods.

**Comment by paramat** [comment]
Maybe better to add the drop on die feature to the bones mod?

File: mods/bones/init.lua [hunk 1]
```diff
@@ -155,21 +161,24 @@ local function may_replace(pos, player)
 	return node_definition.buildable_to and not minetest.is_protected(pos, player:get_player_name())
 end
 
+if minetest.setting_getbool("creative_mode") then
+	return
```
Comment (Ekdohibs) [inline mods/bones/init.lua]: Wouldn't it be better to explicitly disable bones in creative mode with the setting, so players can re-enable it again if they want to?

**Comment by tenplus1** [comment]
@Ekdohibs - That is a decision for server owners and players to make, minetest should only set a default, not make the decision for you based on singleplayer/creative status.

**Comment by t4im** [comment]
Please keep whitespace/code-tidy changes in an extra commit next time. Keeping those apart makes it easier to read the commits and for both git and people to merge them later. And it could prevent overlooking grave errors (sorry for the pun :-D).

**Comment by Ferk** [comment]
In the code you are using the setting "on_die", but in the documentation you wrote the setting is called "bones_enable", that's contradicting. Also you were talking about the minetest.conf setting but not the variable exposed in the bones API  (bones.enable) that mods could use.

I think a better name would be "on_player_death", and I think it should be also bones.on_player_death the API variable, otherwise it's a bit confusing.

Also, maybe there should be a setting by which the items are kept but an (empty) bone node is created. Or a "hardcore" setting where the empty bones are left but the inventory is lost.

**Comment by t4im** [comment]
> variable exposed in the bones API (bones.enable)

That was probably a [missunderstanding](https://github.com/minetest/minetest_game/pull/912).
It won't work like that, because it is always overwritten with `minetest.setting_get("on_die")` before it is being used. So the only way for a mod to change this is `minetest.setting_set("on_die", …)`, which is probably what @tenplus1 wanted anyway.
The `bones.enable` can just as well be removed in favor of already existing minetest API.

**Comment by paramat** [comment]
I feel the game_api.txt needs an example (show the default mode as the example) to show how to format the setting.
I feel the setting name needs to mention bones and be clearer, 'on die' does sound like a function.
Something like 'bones mode', 'bones drop mode', 'bones death mode'.

**Comment by everamzah** [comment]
Is it desirable to have a global variable, `bones`, with a table containing a formspec string, as its only content?  Would it be as good, if not better, to have it as `local formspec` instead?

File: game_api.txt [hunk 2]
```diff
@@ -64,7 +64,7 @@ Beds API
 Bones API
 ---------
 
-The string setting `on_die` will be read from the minetest.conf file ("bones" by default).
+The setting `bones_mode` is read from the minetest.conf file `bones_mode = bones` by default.
```
Comment (paramat) [inline game_api.txt]: Should be
```bones_mode = "bones"```
quotes are needed.
This is a setting not an API so this should be in minetest.conf_example? with the same format: Explanation first then the default setting on a newline after.

**Comment by paramat** [comment]
Thanks.
Added line comment.

**Comment by SmallJoker** [comment]
Came here to make a quick summary (to pull this forwards):
- Bones' behaviour can be changed by the minetest.conf setting `bones_mode` now, this means it's changeable by other mods too (`minetest.setting_set()`)
- `bones` is the default behaviour and works like before
- `drop` drops all items + the bone node
- `keep` will completely disable the bones
- `hybrid` creates bones on unprotected area, otherwise acts like `drop` on protected area

This should include most of the points mentioned above.

**Comment by paramat** [comment]
What does the current code do in othe players' protected areas? Should "hybrid" be the default?

File: mods/bones/init.lua [hunk 3]
```diff
@@ -155,37 +153,69 @@ local function may_replace(pos, player)
...
-	pos.y = math.floor(pos.y+0.5)
-	pos.z = math.floor(pos.z+0.5)
-	local param2 = minetest.dir_to_facedir(player:get_look_dir())
+	local pos = vector.round(player:getpos())
 	local player_name = player:get_player_name()
-	local player_inv = player:get_inventory()
+
+	-- drop inventory and crafting items
+	if bones_mode == "drop" or (bones_mode == "hybrid" and
+			minetest.is_protected(pos, player_name)) then
```
Comment (paramat) [inline mods/bones/init.lua]: Should be using 'may_replace()' instead?

**Comment by paramat** [comment]
Line comment.
"hybrid" is actually very close to current behaviour so seems redundant.
The current behaviour tries to place bones a node higher if protection, but this seems pointless, perhaps we can simplify by removing that feature and making current and "hybrid" identical?

**Comment by tenplus1** [comment]
@paramat - I've removed double drop code and hybrid mode.

File: mods/bones/init.lua [hunk 4]
```diff
@@ -155,37 +153,69 @@ local function may_replace(pos, player)
...
+		obj:setvelocity({
+			x = math.random(-10, 10) / 9,
+			y = 5,
+			z = math.random(-10, 10) / 9,
+		})
+	end
+end
+
 minetest.register_on_dieplayer(function(player)
-	if minetest.setting_getbool("creative_mode") then
```
Comment (paramat) [inline mods/bones/init.lua]: Why was this removed?

File: mods/bones/init.lua [hunk 5]
```diff
@@ -182,63 +182,39 @@ minetest.register_on_dieplayer(function(player)
...
-				minetest.add_item(pos, player_inv:get_stack("craft", i))
-			end
-			-- empty lists main and craft
-			player_inv:set_list("main", {})
-			player_inv:set_list("craft", {})
-			return
-		end
-	end
-	
 	minetest.set_node(pos, {name="bones:bones", param2=param2})
```
Comment (paramat) [inline mods/bones/init.lua]: Spaces needed either side of '='.

**Comment by paramat** [comment]
Added 2 line comments, click 'show 1 comment' above.
Otherwise looks fine, please could you test?

**Comment by tenplus1** [comment]
@paramat - Exactly that, bones drop mode should be determined by the setting alone and not forced into KEEP inventory if creative mode active.  If it's an issue then I'll re-add.

**Comment by paramat** [comment]
:+1: Is it tested?

**Comment by tenplus1** [comment]
@paramat - Is currently running on Xanadu server and working a-ok.

**Comment by paramat** [comment]
Sure, but you need to die in each mode to check it. I'm too busy to check and can't easily set up protected areas.

**Comment by tenplus1** [comment]
@paramat - I've died in each mode inc. creative and it works fine, even as a new player inside a protected area.

**Comment by t4im** [comment]
It doesn't try the node above anymore, does it?
There were some good reasons for doing that.

When digging down in protected areas, client prediction removes the node first, causing the player  to fall into the node, while dying from protection violation inside the dug node.
A common scenario is a guest standing on top of a chest or one-click accidents with some power tools (e.g. technic drill).

Then there are unwalkable nodes, like roofing/carpets or nodes with nodeboxes, like stairs, slabs, slopes, or fences. Of course not all cases were covered before with just checking the node above, but the bones mod got at least a chance for many of the common cases.

**Comment by paramat** [comment]
t4im i didn't realise how useful that was, ok i think that needs to be re-added.

**Comment by tenplus1** [comment]
Placing a single node higher doesn't always work remember, what if you are moving 1 node TOWARDS the protection, and if not would 1 node really help that much ?!?! NO!  If a player dies inside of a protected area then current behaviour will DROP the items instead of dropping a bone.

**Comment by paramat** [comment]
:+1: Let's merge this and someone else can improve it.

**Comment by t4im** [comment]
> what if you are moving 1 node TOWARDS the protection

Before it would only place one node higher, if the current position wouldn't work.  So chances weren't decreased. Bones were given a second chance, which helped with fences, slopes, stairs, horizontal slabs, doormats, wheat, cotton and so on.

**Comment by tenplus1** [comment]
Current operation means if bones cannot be placed then items are dropped, so the failsafe is already included.

**Comment by Ferk** [comment]
> if bones cannot be placed then items are dropped

With one itemstack entity per inventory slot? couldn't that potentially cause trouble? Someone could make a massmurder area with lava that could potentially result in too many entities to handle, wouldn't it?

**Comment by paramat** [comment]
Dropping items on stairs, slabs, carpet etc will be undesirable, so i think an attempt should be made to place bones one node higher, perhaps this should be seperated from the protection check.
Anyway since the author is not happy with doing this shall we merge this and someone else can complete it?

**Comment by paramat** [comment]
:+1:

**Comment by t4im** [comment]
That should work :+1:, thanks.

> The comment misled me into thinking the extra check was for this only.

Yea, sorry about that. My code-comment wasn't very well phrased.
I focused on it, because the "dig straight down" scenario was also replacing those chests with bones (gravity pulled them into that position before the dig was reverted). The rest seemed minor compared to that. :-/

**Comment by paramat** [comment]
@sofar please can you review?

**Comment by everamzah** [comment]
This makes items drop to the ground in protected areas by default?  That is a change in behavior, to be sure.  :-1:

**Comment by paramat** [comment]
Nope that was previous behaviour too.

**Comment by everamzah** [comment]
Excuse me?  I have always played with bones, since version 0.4.9, and in each instance, the /only/ time items dropped, is if there was not enough room to fit everything, spilling the remainder.  I suggest you fire up a stock minetest_game instance and give this one a double-check.

The only thing that currently happens in protected areas is that the bones node goes fresher faster (time / 4).

**Comment by paramat** [comment]
I'm probably wrong, i'll check the code.

**Comment by paramat** [comment]
It was https://github.com/minetest/minetest_game/commit/75ea7e3160321fd60c80907accb427650391bb0e
Protected nodes should never be replaced, even by bones.

**Comment by everamzah** [comment]
I can only speak from in-game experience, where I am typically in sync with HEAD.  Yes, if a bones node is going to try to replace a protector node, it will try to put it one higher, and if that fails, perhaps drop the nodes.  I have seen a lot of bones, but cannot recall a single time a player's inventory fell immediately.

If items drop in protected areas, all Player A needs to do is place a protector node before felling Player B to steal their items!  There goes the five minute grace period for Player A to get their stuff back... That's a pretty radical change, I'd say.

**Comment by paramat** [comment]
^ @t4im thoughts on this?

**Comment by paramat** [comment]
@everamzah if this is not done in this PR please can you open an issue to remind us?

**Comment by t4im** [comment]
`may_replace()` does the protection check last (for known nodes). It returns `true` if it's air or liquid, even if protected.

That's why the previous check was ok with either y or y+1 being just protected air.

I'm not quite sure yet why this version of it wouldn't do that yet.
Time to test :-)

**Comment by paramat** [comment]
Earlier i wrote:

> "hybrid" is actually very close to current behaviour so seems redundant.
[..] perhaps we can [..] [make] current [behaviour] and "hybrid" identical?

So this may be partly my fault through misunderstanding the behaviour.

**Comment by everamzah** [comment]
I'm always open to creating new issues.

**Comment by paramat** [comment]
> may_replace() does the protection check last (for known nodes). It returns true if it's air or liquid, even if protected.

Yes this is wrong. If the node is air (very likely) then there is no protection check at all.

**Comment by paramat** [comment]
I misunderstood, 'may replace' does not do the protection check, if the pos is air (very likely) then it returns true.

**Comment by paramat** [comment]
Looking again, current behaviour seems fine. Now i'll re-consider this PR.

**Comment by paramat** [comment]
:+1:

**Comment by everamzah** [comment]
I am compiling remotely now, and will test it all out.  Poor TenPlus1!

**Comment by paramat** [comment]
Unfortunately i don't fully understand every aspect of MTGame, but slowly improving. Apologies to tenplus1.

**Comment by everamzah** [comment]
Having just tested it, my concerns are lifted as Player B has safe, fresh bones protected against Player A's protection, as it is in HEAD.  Now I'm waiting for 5 to 20 minutes to see what happens.  :+1:

**Comment by paramat** [comment]
Thanks.

**Comment by paramat** [comment]
Approval from sfan5 http://irc.minetest.ru/minetest-dev/2016-07-25#i_4656720

**Comment by paramat** [comment]
error: patch failed: mods/bones/init.lua:155
error: mods/bones/init.lua: patch does not apply

This may need rebasing.

**Comment by SmallJoker** [comment]
@paramat There were no conflicts. However, does it work now?

**Comment by paramat** [comment]
Sometimes the message in the thread is wrong. Will try again.

**Comment by paramat** [comment]
2df7ce20dd8290b677cc32c4b4a1bde3654c137f
Congratulations!
